pkg/prometheus/watcher.go

```diff
@@ -470,3 +605,105 @@ func (w *Watcher) syncAlertManager(key string) error {
 	_, err = w.mClient.MonitoringV1().Alertmanagers(am.Namespace).Update(am)
 	return err
 }
+
+func (w *Watcher) handleAlertmanagerConfigAdd(obj interface{}) {
+	key, ok := keyFunc(obj)
+	if !ok {
+		return
+	}
+
+	log.Debugf("Alertmanager config added: %s", key)
+
+	enqueue(w.amcQueue, key)
+}
+
+func (w *Watcher) handleAlertmanagerConfigUpdate(oldObj, newObj interface{}) {
+	key, ok := keyFunc(newObj)
+	if !ok {
+		return
+	}
+
+	log.Debugf("Alertmanager config added: %s", key)
+
+	enqueue(w.amcQueue, key)
+}
+
+func (w *Watcher) handleAlertmanagerConfigDelete(obj interface{}) {
+	key, ok := keyFunc(obj)
+	if !ok {
+		return
+	}
+
+	log.Debugf("Alertmanager config deleted: %s", key)
+
+	enqueue(w.amcQueue, key)
+}
+
+func (w *Watcher) syncAlertManagerConfig(key string) error {
+	var alertManagerName string
+	var data []byte
+
+	log.Debugf("syncing alert manager config: key: %s", key)
+	obj, exists, err := w.amcInf.GetIndexer().GetByKey(key)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		log.Error("Key not found...")
+		return nil
+	}
+
+	amc := obj.(*monitoringv1.AlertmanagerConfig)
+	if nil == amc {
+		log.Errorf("Got an invalid amc object for: %s", key)
+		return nil
```

Comment:
seems like coding bug. We should assert here, not just let this go.

Reply:
We also get objects in syncAlertmanager and syncPrometheus function, but there was not check there. Since this is a long running loop, if we assert it will exit and the pod will restart. If the error condition prevails the pod keeps on restarting. 

returning nil also causes an immediate call back to retry at least a few time before it gives up.

Reply:
ok